File: ios/Sources/Shared/PostsRepository.swift
import Foundation

/// posts.json을 GitHub raw URL에서 내려받아 App Group 공유 컨테이너에 캐시한다.
/// 앱과 위젯 익스텐션이 동일한 캐시를 읽고 쓰기 때문에, 둘 중 하나가 최근에
/// 갱신했다면 나머지는 네트워크 없이 그 결과를 바로 쓸 수 있다.
final class PostsRepository {
    static let shared = PostsRepository()

    /// Xcode "Signing & Capabilities > App Groups"에서 앱/위젯 두 타겟 모두에
    /// 등록한 값과 반드시 동일해야 한다. (기본값은 project.yml의 entitlements와 일치)
    static let appGroupID = "group.com.wooyxxng.pptnzblog"

    /// WidgetKit이 타임라인을 갱신할 때 쓰는 위젯 kind. 위젯 정의와 문자열이 반드시 일치해야 한다.
    static let widgetKind = "PPTNZBlogWidget"

    /// crawl.yml이 posts.json을 main 브랜치에 커밋하므로 그걸 그대로 읽는다.
    private let remoteURL = URL(
        string: "https://raw.githubusercontent.com/wooyxxng-Jang/pptnzblog/main/posts.json"
    )!

    private var cacheFileURL: URL? {
        FileManager.default
            .containerURL(forSecurityApplicationGroupIdentifier: Self.appGroupID)?
            .appendingPathComponent("posts_cache.json")
    }

    /// 네트워크 우선, 실패하면 캐시로 폴백. 캐시도 없으면 에러를 던진다.
    func loadPosts() async throws -> [Post] {
        do {
            let (data, _) = try await URLSession.shared.data(from: remoteURL)
            let posts = try JSONDecoder().decode([Post].self, from: data)
            writeCache(data)
            return posts
        } catch {
            if let cached = readCache() {
                return cached
            }
            throw error
        }
    }

    /// 네트워크 왕복 없이 즉시 뭔가 보여줘야 하는 위젯 placeholder/snapshot용.
    func loadCachedPosts() -> [Post] {
        readCache() ?? []
    }

    private func readCache() -> [Post]? {
        guard let url = cacheFileURL,
              let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode([Post].self, from: data)
    }

    private func writeCache(_ data: Data) {
        guard let url = cacheFileURL else { return }
        try? data.write(to: url, options: .atomic)
    }
}

File: android/app/src/main/java/com/wooyxxng/pptnzblog/data/PostsRepository.kt
package com.wooyxxng.pptnzblog.data

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException

/**
 * posts.json을 GitHub raw URL에서 내려받아 DataStore에 캐시한다.
 * crawl.yml이 posts.json을 main 브랜치에 커밋하므로 그걸 그대로 읽는다.
 */
class PostsRepository(context: Context) {
    private val preferences = AppPreferences(context.applicationContext)
    private val client = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }

    companion object {
        /** Glance 위젯이 갱신을 트리거할 때 쓰는 식별자. */
        const val WIDGET_KIND = "PptnzWidget"

        private const val REMOTE_URL =
            "https://raw.githubusercontent.com/wooyxxng-Jang/pptnzblog/main/posts.json"

        @Volatile private var instance: PostsRepository? = null

        fun getInstance(context: Context): PostsRepository =
            instance ?: synchronized(this) {
                instance ?: PostsRepository(context).also { instance = it }
            }
    }

    /** 네트워크 우선, 실패하면 캐시로 폴백. 캐시도 없으면 에러를 던진다. */
    suspend fun loadPosts(): List<Post> = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder().url(REMOTE_URL).build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) throw IOException("HTTP ${response.code}")
                val body = response.body?.string() ?: throw IOException("empty body")
                val posts = json.decodeFromString<List<Post>>(body)
                preferences.savePostsCache(body)
                posts
            }
        } catch (e: Exception) {
            val cached = loadCachedPosts()
            if (cached.isNotEmpty()) cached else throw e
        }
    }

    /** 네트워크 왕복 없이 즉시 뭔가 보여줘야 하는 위젯 placeholder/snapshot용. */
    suspend fun loadCachedPosts(): List<Post> = withContext(Dispatchers.IO) {
        val raw = preferences.loadPostsCache() ?: return@withContext emptyList()
        runCatching { json.decodeFromString<List<Post>>(raw) }.getOrDefault(emptyList())
    }
}
